---
title: "Supplemental Figure - Generate heatmap of pairwise genotype-phenotype transition data across reconstructions"
author: "Kyle Gontjes"
date: "`r Sys.Date()`"
output: html_document
---

```{r setup, include=FALSE}
require(knitr) 
opts_knit$set(echo = TRUE,warning = FALSE, message = FALSE,error=FALSE,comment=NA) 
opts_chunk$set(echo = TRUE,warning = FALSE, message = FALSE,error=FALSE,comment=NA) 
knitr::opts_knit$set(root.dir = '~/Desktop/gl_mount/Analysis/Phyloaware/joint-uncertainty-ms/') 
```

# Environment
```{r environment, echo=T, message=FALSE, include=T, results=T}
#Packages
packages <- c("tidyverse",'kableExtra','ComplexHeatmap','RColorBrewer','colorRamp2') 

#Load Packages
invisible(lapply(packages,library,character.only=T,quietly=T))

# Source functions
source("./lib/consistent_themes.R")

# Print environment
set.seed(0104) 
Sys.info()
sessionInfo()
```
 
# Load data  
## Reconsctructions
```{r, reconstructions}  
clustering_str <- readRDS("./data/empirical/joint_reconstruction/hclust_consensus_distance_matrix_seven_clusters_string.RDS")
```

## Pairwise phyloAMR data
```{r, pairwise data}
pairwise_comparisons <- readRDS("./data/empirical/phyloAMR/pairwise_synchronous_transitions_genotype_as_comparitor.RDS")  %>% subset(genotype_reconstruction != 'focal_states') %>% subset(genotype %in% c("AA552","OmpK36_c25t","OmpK36GD","OmpK36TD","OmpK36_promoter_IS","OmpK36_disruptive","PBP_any","RamR","RamA"))
```

## Recode genotypes
```{r, recode genotypes}
# Recode
pairwise_comparisons$genotype <- factor_genotypes(pairwise_comparisons$genotype) 
pairwise_comparisons$genotype <- recode_genotypes(pairwise_comparisons$genotype) 
levels(pairwise_comparisons$genotype) <- collapse_PFAV(levels(pairwise_comparisons$genotype))

# Order
pairwise_comparisons_ordered <- pairwise_comparisons %>% arrange(genotype,pheno_cluster)
pairwise_comparisons_ordered$genotype_reconstruction_axes <- paste0(pairwise_comparisons_ordered$genotype,"_",pairwise_comparisons_ordered$genotype_reconstruction)
```

# Synchronous gain events
```{r, sync, fig.width=8.5, fig.height=6.8, results='hide',echo=TRUE}
# Pivot wide
matrix_of_vals <- pairwise_comparisons_ordered %>% arrange(genotype) %>% 
    group_by(genotype_reconstruction_axes, phenotype_reconstruction)  %>% 
    summarise(value = synchronous_gains_num, genotype = genotype) %>%  arrange(genotype) %>% select(-genotype) %>% 
    pivot_wider(names_from = genotype_reconstruction_axes, values_from = value)  

# Get cluster calls
cluster_names <- clustering_str %>% `names<-`(seq_len(length(.)))
matrix_of_vals$cluster <- sapply(matrix_of_vals$phenotype_reconstruction,FUN=function(x){
  if(!grepl("focal",x)){
    cluster_names[names(cluster_names) == gsub("alt_recon_","",x)]
  } else {
    x
  }
})

matrix_of_vals$cluster <- recode_phenotype_clusters(matrix_of_vals$cluster)

matrix_of_vals <- matrix_of_vals %>% arrange(cluster) 

# Row data
rowsplit_pheno <- factor(matrix_of_vals$cluster,levels = unique(sort(matrix_of_vals$cluster)))

gene <- factor(colnames(matrix_of_vals %>% select(-cluster, -phenotype_reconstruction)) %>% str_split(.,"_",simplify=T) %>% .[,1:2] %>% apply(.,1,paste0,collapse="_")  %>% gsub("_alt|_focal","",.),levels = levels(pairwise_comparisons_ordered$genotype)) 

# Colors & annotations
phen_colors <- setNames(phylogenetic_reconstruction_scale$palette(length(phylogenetic_reconstruction_scale$breaks)),phylogenetic_reconstruction_scale$breaks)

## Row annotations
rowannots <- ComplexHeatmap::rowAnnotation(
  `Reconstruction axis` = factor(matrix_of_vals$cluster),
  annotation_name_gp = gpar(fontsize = 0),
  col = list(`Reconstruction axis` = phen_colors),
    show_legend = FALSE 
)

## Column annotations
genotype_levels <- levels(pairwise_comparisons$genotype) 
genotype_colors <- setNames(hues::iwanthue(length(genotype_levels)), genotype_levels)

colannots <- ComplexHeatmap::columnAnnotation(
  `Genotype` = gene,
  annotation_name_gp = gpar(fontsize = 0),
  col = list(`Genotype` = genotype_colors),
   show_legend = FALSE 
) 

max_syn_transitions <- matrix_of_vals %>% select(-cluster, -phenotype_reconstruction) %>% as.matrix() %>% unlist %>% as.vector %>% max

col_fun_syn <- colorRamp2(c(0, max_syn_transitions), c("white", "red"))

A <- Heatmap(
  matrix_of_vals %>% select(-cluster, -phenotype_reconstruction) %>% as.matrix(),
  row_split = rowsplit_pheno,
  column_split = gene,
  cluster_rows = FALSE,
  cluster_row_slices = TRUE,
  cluster_columns = FALSE,
  cluster_column_slices = TRUE,
  col =col_fun_syn,
  show_row_names = FALSE,
  show_column_names = FALSE,
  show_row_dend = FALSE,
  row_title = NULL,
  column_title = NULL,
  right_annotation = rowannots,
  top_annotation = colannots,
  name = "Synchronous gain events",
  heatmap_legend_param =list(at = 0:max_syn_transitions,labels = 0:max_syn_transitions),
  use_raster = F
)

A

# Extract legends for each annotation
lgd_row <- color_mapping_legend(rowannots@anno_list[["Reconstruction axis"]]@color_mapping)
lgd_col <- color_mapping_legend(colannots@anno_list[["Genotype"]]@color_mapping)

# Pack all legends into one column
merged_legends <- packLegend( 
  lgd_row,
  lgd_col,
  direction = "vertical"
)
```

# Synchronous loss events
```{r, loss, fig.width=8.5, fig.height=6.8, results='hide', echo=TRUE}
# Pivot wide 
matrix_of_vals_loss <- pairwise_comparisons_ordered %>% arrange(genotype) %>%
    group_by(genotype_reconstruction_axes, phenotype_reconstruction) %>% 
    summarise(value = synchronous_losses_num,genotype=genotype) %>%  arrange(genotype) %>% select(-genotype)  %>% 
    pivot_wider(names_from = genotype_reconstruction_axes, values_from = value)  

matrix_of_vals_loss$cluster <- sapply(matrix_of_vals_loss$phenotype_reconstruction,FUN=function(x){
  if(!grepl("focal",x)){
    cluster_names[names(cluster_names) == gsub("alt_recon_","",x)]
  } else {
    x
  }
})

matrix_of_vals_loss$cluster <- recode_phenotype_clusters(matrix_of_vals_loss$cluster)

matrix_of_vals_loss <- matrix_of_vals_loss %>% arrange(cluster) 
rowsplit_pheno <- factor(matrix_of_vals_loss$cluster,levels = unique(sort(matrix_of_vals_loss$cluster)))

gene <- factor(colnames(matrix_of_vals_loss %>% select(-cluster, -phenotype_reconstruction)) %>% str_split(.,"_",simplify=T) %>% .[,1:2] %>% apply(.,1,paste0,collapse="_")  %>% gsub("_alt|_focal","",.),levels = levels(pairwise_comparisons_ordered$genotype)) 

# Colors and annotations 
max_loss_transitions <- matrix_of_vals_loss %>% select(-cluster, -phenotype_reconstruction) %>% as.matrix() %>% unlist %>% as.vector %>% max

col_fun_loss <- colorRamp2(c(0, max_loss_transitions), c("white", "red"))

## Row annotations
rowannots <- ComplexHeatmap::rowAnnotation(
  `Reconstruction axis` = factor(matrix_of_vals_loss$cluster),
  annotation_name_gp = gpar(fontsize = 0),
  col = list(`Reconstruction axis` = phen_colors),
    show_legend = FALSE 
)

## Column annotations
colannots <- ComplexHeatmap::columnAnnotation(
  `Genotype` = gene,
  annotation_name_gp = gpar(fontsize = 0),
  col = list(`Genotype` = genotype_colors),
   show_legend = FALSE 
) 

B <- Heatmap(
  matrix_of_vals_loss %>% select(-cluster, -phenotype_reconstruction) %>% as.matrix(),
  row_split = rowsplit_pheno,
  column_split = gene,
  cluster_rows = FALSE,
  cluster_row_slices = TRUE,
  cluster_columns = FALSE,
  cluster_column_slices = TRUE,
  col = col_fun_loss,
  show_row_names = FALSE,
  show_column_names = FALSE,
  show_row_dend = FALSE,
  row_title = NULL,
  column_title = NULL,
  right_annotation = rowannots,
  top_annotation = colannots,
  name = "Synchronous loss events",
  heatmap_legend_param =list(at = 0:max_loss_transitions,labels = 0:max_loss_transitions),
  use_raster=F
)

B

# Extract legends for each annotation
lgd_row <- color_mapping_legend(rowannots@anno_list[["Reconstruction axis"]]@color_mapping)
lgd_col <- color_mapping_legend(colannots@anno_list[["Genotype"]]@color_mapping)

# Pack all legends into one column
merged_legends_2 <- packLegend( 
  lgd_row,
  lgd_col,
  direction = "vertical"
)
``` 

# Save figures
```{r, save figures}
png("./figures/empirical/pairwise_reconstruction_synchronous_transition_comparisons_synchronous_gain_events_heatmap.png", width=8.5, height=6.8, units="in", res=900)
draw(
  A,
  heatmap_legend_list = list(merged_legends),
  annotation_legend_list = NULL
)
dev.off()

png("./figures/empirical/pairwise_reconstruction_synchronous_transition_comparisons_synchronous_loss_events_heatmap.png", width=8.5, height=6.8, units="in", res=900)
draw(
  B,
  heatmap_legend_list = list(merged_legends),
  annotation_legend_list = NULL
)
dev.off()
```
